Face enhancements, multiline verbatims

cdeleuze opened the pull request:
- Ability to specify a face for the question heading (might be different from the rest of the question text face).
- Handle multiline verbatim is consistent with recent commit #7b0d4b5a in auto-multiple-choice.
[commit on gitlab](https://gitlab.com/jojo_boulix/auto-multiple-choice/-/commit/7b0d4b5a4580bd21b0310add5eed4db59edd3190)

wentasah commented:
Thank you very much for the PR. It looks nice, but there are some minor glitches.

The first commit looks innocent, but it breaks highlighting of groups. I tried to investigate the reason, but I have no clue why it is so. Maybe you'd have some idea. I'm testing it with the following file:
```
# -*- amc-txt -*-

*( Group

* Question heading
Question content
+ Answer

*) End of group
```
In the master branch the group lines are highlighted as dark gray. When I add your first commit, the lines turn into `amc-txt-question-heading`.

The second commits seems OK. I'd appreciate if the [example](https://gitlab.com/jojo_boulix/auto-multiple-choice/uploads/368c0fc17bd5a531f869191b3bc2a377/source.txt) could be a part of the commit message.

cdeleuze commented:
Oh, I'm not using groups much and it seems I've been a bit floppy on the face thing.  I think the problem comes from the use of the override flag in the `amc-txt-question-heading` entry in the font-lock keywords, combined with the fact that group lines are matched *also* by `amc-txt-question-re`. I'll give it a closer look soon and will amend the commits if I can solve this satisfactorily.

cdeleuze committed:
Face for question heading

cdeleuze committed:
Handle multiline verbatims (see amc commit #7b0d4b5a)

Eg in the following example, the directory listing lines inside
the verbatim must not be fontified as negative answers.

Title: Verbatim example

* Lulu's home dir

[verbatim]$ pwd
/home/lulu
$ ls -l
-rw-r--r--  1 lulu lulu    42192 Feb  4 16:32  titi.png
-rw-r--r--  1 lulu lulu      191 Feb  4 16:29  toto.txt
[/verbatim]
Lulu's home dir contains
  - exactly two files
  + two files or more

cdeleuze commented:
I changed `amc-txt-question-re` to not match group lines.  Unfortunately AMC-TXT does not force a space after the leading `*` or `**` of question heading and that makes it a bit tricky.  The new definition is not strictly correct (see comment in source) but this will show up only in very pathological cases.  Any feedback welcome.

wentasah commented on amc-txt.el:
```diff
@@ -41,20 +41,46 @@
+      ;; option is included, eg in `*[horiz](Question' group 5 will
+      ;; include the [horiz] part.  We may consider this a
+      ;; pathological case.
+      (* blank) (group-n 5 (not (any "(" ")")) (* any))))
```
I have tried to implement this differently. See [this commit](https://github.com/wentasah/amc-txt-mode/commit/8659ce9cc1663189ee28fc58ef9d973851bfdf6f) (not yet in master). The rationale it that when we already use the `amc-txt-search-question` function rather than just the plain regular expression `amc-txt-search-question-re`, we can extend that function to achieve exactly what we need, without considering a pathological case.

cdeleuze replied:
seems ok

wentasah commented on amc-txt.el:
```diff
@@ -41,20 +41,46 @@
+;;   - add amc-txt-font-lock-extend-verbatim to
+;;     font-lock-extend-region-functions.
+;;
+;; There's still the problem of amc-txt-question-re being stopped too
+;; early at an answer-like pattern inside a verbatim.
```
This problem is still there, but it could also be later resolved by extending `amc-txt-search-question`.

Also see [this commit](https://github.com/wentasah/amc-txt-mode/commit/2818234e27e058c731a9b35c014358298d990f57), which helps in rendering verbatim content with lines starting with `*` (I'm not sure whether AMC handles such things correctly, though).

cdeleuze replied:
AMC does handle * in verbatim correctly yes. It reads input line by line and if `[verbatim]` is found it eats following lines until `[/verbatim]` without considering other regexes.

wentasah commented on amc-txt.el:
```diff
@@ -112,6 +138,14 @@ Value nil is the same as 1."
+(defun amc-txt-font-lock-extend-verbatim ()
+  "Move fontification boundaries for verbatim construct."
+  (goto-char font-lock-beg)
+  (when (equal (get-text-property (point) 'face) 'amc-txt-verbatim)
```
It seems to me that this doesn't work well. Probably, we should extend the region not only when `font-lock-beg` lies in the verbatim block, but also when `font-lock-end` lies there. Also, I'm not sure whether relying on `face` inside font-lock code is reliable. When changing things inside verbatim (such as removing and adding `-` at the start of the line), rehighliting is broken. My attempt to fix all these things is [here](https://github.com/wentasah/amc-txt-mode/commit/3e87b510efa32ee091aa67b634e0bb81218cd8f9), but it's not perfect. Currently, I think that solving rehighlighting correctly will always require to search for `[verbatim]` and `[/verbatim]`, without relying on text properties (face or something else).

wentasah reviewed (commented):
I've played with the code a bit and it looks like a can of worms :-) There is always something slightly broken. I tried to fix one thing and another thing appears. Now, it's too late and I'm still not done. So I've pushed my current state to [font-lock-updates](https://github.com/wentasah/amc-txt-mode/commits/font-lock-updates) branch. Feedback welcome too. I'll try to continue in upcoming days.
